File: unittests/ym/common/random/testsuite.py
##
# @file    testsuite.py
# @version 1.0.0
# @author  Forrest Jablonski
#

import os
import sys

try:
   # @note Grabs the first directory in the chain named unittests/.
   sys.path.append(os.path.join(os.getcwd().split("unittests")[0], "unittests/"))
   import testsuitebase
except:
   print("Cannot import testsuitebase - path set correctly?")
   sys.exit(1)

try:
   import cppyy

   from cppyy.gbl import std
   from cppyy.gbl import ym
except:
   print("Cannot import cppyy - started the venv?")
   sys.exit(1)

class TestSuite(testsuitebase.TestSuiteBase):
   """
   @brief Collection of all tests for SUT Random.
   """

   def __init__(self):
      """
      @brief Constructor.
      """
      super().__init__("ym/common/", "random", "Random", [])

      self.test_cases = {
         "ZerosAndOnes": self.analyze_results_ZerosAndOnes
      }

   def run(self, test_case_names:list=None):
      """
      @brief Runs the test suite or set of test cases.
      """

      ts = ym.ut.TestSuite()

      if test_case_names is not None:
         for name in test_case_names:
            try:
               results = ts.runTestCase(name)
               self.test_cases[name](results)
            except:
               print(f"Something went wrong running test case {name}")
      else:
         all_results = ts.runAllTestCases()
         for name in all_results:
            self.test_cases[name](all_results[name])

   def analyze_results_ZerosAndOnes(self, results):
      """
      @brief Analyzes results from the zeros and ones test.
      """

      nTotalBits = results.get[ym.ut.utuint64]("NTotalBits", 0)
      nSetBits   = results.get[ym.ut.utuint64]("NSetBits", 0)
      print(f"nTotalBits = {nTotalBits}")
      print(f"nSetBits = {nSetBits}")
      print(f"Ratio = {nSetBits/nTotalBits}")

# kick-off
if __name__ == "__main__":
   if os.path.basename(os.getcwd()) != "random":
      print("Needs to be run in the random/ directory")
      sys.exit(1)

   ts = TestSuite()
   ts.run()

File: unittests/ym/common/random/testsuite.cpp
/**
 * @file    testsuite.cpp
 * @version 1.0.0
 * @author  Forrest Jablonski
 */

#include "ymdefs.h"
#include "testsuite.h"

#include "random.h"

#include <bitset>

/** TestSuite
 *
 * @brief Constructor.
 */
ym::ut::TestSuite::TestSuite(void)
   : TestSuiteBase("Random")
{
   addTestCase<ZerosAndOnes>();
}

/** run
 *
 * @brief Runs zeros and ones test.
 * 
 * @note Amount of zeros and ones should be about equal.
 *
 * @returns DataShuttle -- Important values acquired during run of test case.
 */
auto ym::ut::TestSuite::ZerosAndOnes::run([[maybe_unused]] DataShuttle const & InData) -> DataShuttle
{
   ym::Random rand;

   constexpr auto NIters     = 1'000'000'000_u64;
   constexpr auto NTotalBits = NIters * 64_u64;
             auto nSetBits   = 0_u64;

   for (auto j = 0_u64; j < NIters; ++j)
   {
      nSetBits += std::bitset<64>(rand.gen<uint64>()).count();
   }

   return {
      {"NTotalBits", NTotalBits},
      {"NSetBits",   nSetBits}
   };
}

// /** run
//  *
//  * @brief Runs uniform bins test.
//  * 
//  * @note Amount of entries in each bin should be about equal.
//  *
//  * @returns bool -- True if test passed, false otherwise.
//  */
// auto ym::ut::Random_UT::UniformBins_TC::run(void) -> std::unordered_map<std::string, double>
// {
//    ym::Random rand;

//    std::vector histo(1_u64 << 16_u64, 0_u64);

//    constexpr auto NIters = 1_u64 << 30_u64;

//    for (auto i = 0_u64; i < NIters; ++i)
//    {
//       auto const R = rand.gen<uint64>();

//       histo[(R >> 48_u64) & 0xffff_u64] += 1_u64;
//       histo[(R >> 32_u64) & 0xffff_u64] += 1_u64;
//       histo[(R >> 16_u64) & 0xffff_u64] += 1_u64;
//       histo[(R >>  0_u64) & 0xffff_u64] += 1_u64;
//    }

//    auto const NExpectedHitsPerBin = NIters / histo.size();

//    return std::unordered_map<std::string, float64>(); // TODO
// }

// /** run
//  *
//  * @brief Runs the bin frequency test.
//  * 
//  * @todo I want to use ym::MemMan, add dependency for that test to run first.
//  *
//  * @returns bool -- True if test passed, false otherwise.
//  */
// auto ym::ut::Random_UT::BinFrequency_TC::run(void) -> std::unordered_map<std::string, double>
// {
//    ym::Random rand;

//    return std::unordered_map<std::string, double>();
// }
